### PR TITLE
[Frontend][NFC] Refactor validateTBDIfNeeded

### PR DESCRIPTION
**Pull request by CodaFi**
Inline the conditions leading to `astGuaranteedToCorrespondToSIL` and clean up the logic here to be more const-friendly. Also try to centralize and document the reasons we skip TBD validation.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by compnerd** [comment]
@CodaFi Linux has been busted for ~36 hours.  Looking to see if #31144 helps

**Comment by CodaFi** [comment]
@swift-ci test Linux

File: lib/FrontendTool/FrontendTool.cpp [hunk 1]
```diff
@@ -1422,44 +1413,76 @@ static bool processCommandLineAndRunImmediately(const CompilerInvocation &Invoca
...
+    switch (mode) {
+    case FrontendOptions::TBDValidationMode::None:
+      llvm_unreachable("Handled Above!");
+    case FrontendOptions::TBDValidationMode::Default:
+    case FrontendOptions::TBDValidationMode::MissingFromTBD:
+      return false;
+    case FrontendOptions::TBDValidationMode::All:
+      return true;
+    }
+  }();
```
Comment (davidungar) [inline lib/FrontendTool/FrontendTool.cpp:1474]: Nice construction allowing for a `const`.

**Review by davidungar — APPROVED** [review]
Nice!

**Comment by CodaFi** [comment]
⛵

**Pull request merged**